### DISCUSSION TITLE
V7 nested MantineProvider to force dark theme on partial UI

### DISCUSSION THREAD

**Discussion by naivefun** [mantinedev/mantine#4803] [Q&A]
In V6 we used nested `MantineProvider` to force partial of the UI to always be with dark mode. For example, the header will always remain dark no matter the app theme is dark or light. It is convenient because all components within the header will adapt automatically.

In V7 if I use force color scheme, it will overwrite app level color scheme, causing the whole app always to be dark.

Is there a way to force partial of the UI to always be dark without affecting the app theme?

**Comment by rtivital** [comment]
Use nested MantineProvider:
- Update css variables selector – https://mantine.dev/theming/mantine-provider/#cssvariablesselector
- Change root element – https://mantine.dev/theming/mantine-provider/#getrootelement
- Add custom color scheme manager that updates the attribute – https://mantine.dev/theming/mantine-provider/#colorschememanager

**Reply by naivefun** [reply]
Nested provider can force dark mode. However it seems also affect root level provider's color scheme.

**Reply by rtivital** [reply]
You need to change root element – https://mantine.dev/theming/mantine-provider/#getrootelement

**Reply by Toshinaki** [reply]
Hi, many thanks for the v7. All updates look great.
I remember there're examples for nested providers to use didferent theme for part of the app in the previous version's doc.
Could you please add a similar example to the newest doc?

**Reply by Toshinaki** [reply]
@naivefun 
Hi, did you find any solution to this?

@rtivital 
if I understand it right, we need to:
```tsx
<MantineProvider> // this will use `html` and default storage key
  <div>
    <MantineProvider 
      getRootElement={() = > document.getElementById("section1")} 
      cssVariablesSelector="#section1" 
      colorSchemeManager={localStorageColorSchemeManager({ key: 'section1-scheme' })}
    >
      <div id="section1"></div>
    </MantineProvider>
    <MantineProvider 
      getRootElement={() = > document.getElementById("section2")} 
      cssVariablesSelector="#section2" 
      colorSchemeManager={localStorageColorSchemeManager({ key: 'section2-scheme' })}
    >
      <div id="section2"></div>
    </MantineProvider>
  </div>
</MantineProvider>
```
Is this code right?

**Reply by alexbatis** [reply]
hi @rtivital  - I used [this nextjs app dir mantine template](https://github.com/mantinedev/next-app-min-template) to generate a reproduction  ([code sandbox](https://codesandbox.io/p/github/alexbatis/mantine-nested-color-scheme-reproduction/master)/[github repo](https://github.com/alexbatis/mantine-nested-color-scheme-reproduction)) to outline some issues I've been facing with **nested Mantine theme providers**.  These examples build off of the [example codeblock](https://github.com/orgs/mantinedev/discussions/4803#discussioncomment-7660910) from @Toshinaki above.

_note: this seemed like an appropriate place to report my findings, but let me know if i should add this as an issue on the mantine repo._ 

I'm looking forward to resolving these issues with you


issue 1
- dark theme inside of light theme works, light theme nested inside of dark theme does not (different behaviors between `Card` and `Paper` component)

https://github.com/mantinedev/mantine/assets/25943611/1aae0c68-a6dc-4895-8ac8-0b812ff64159


issue 2 
- nested toggle color scheme not working first time after parent color scheme is changed

https://github.com/mantinedev/mantine/assets/25943611/360c3e6b-d1a6-443c-815b-95fc56b789ae


issue 3 
- `forceColorScheme` not working for nested `MantineProvider`s
<img width="2168" alt="Screenshot 2023-11-25 at 6 22 26 PM" src="https://github.com/mantinedev/mantine/assets/25943611/ccf20634-937d-4c9e-90ff-4dde3cb41b01">

issue 4
- custom color passed to nested mantine provider not working on first render, shows up after calling `toggleColorTheme` in the nested provider

https://github.com/mantinedev/mantine/assets/25943611/070c118c-f4d0-43c3-a236-156a2ca43497


issue 5 (perhaps not an issue, but misunderstanding)
- local storage seems to work when refreshing page, but `useMantineColorScheme` seems to return the wrong value 

https://github.com/mantinedev/mantine/assets/25943611/42cb930d-34cb-4a3d-b388-d41fb96cd953

**Reply by rtivital** [reply]
Thanks for the reproduction! It is planned to add documentation for nested MantineProviders in one of the next minor releases. I'll have a look at these bugs as well.

**Reply by kevinbeal** [reply]
If you use nested mantine providers with different scheme values, the `@mixin light` and `@mixin dark` styles both apply. Whichever (light or dark) is defined last wins.

**Comment by alexturpin** [comment]
The following seems to have worked for me:

```
<MantineProvider
  getRootElement={() => document.querySelector<HTMLElement>("#dark-area") ?? undefined}
  cssVariablesSelector="#dark-area"
  forceColorScheme="dark"
>
  {children}
</MantineProvider>
```

With the caveat that text elements were still getting their `color` from the body. Throwing this in my CSS modules did the trick:

```
:global {
  #aside {
    color: var(--mantine-color-text);
  }
}
```

**Reply by ihqtim** [reply]
Thanks! This is simple to apply & works well for me for most children of the nested provider.  Two issues I see are that the  background color of the TagsInput - in this case the top-most provider is dark theme, and this nested area is light theme:
![image](https://github.com/mantinedev/mantine/assets/115905886/3d7ad11c-b1e2-4730-be84-fa59d20436ba)

Similarly Select components from the child (nested) provider have their dropdown coloring set by the top level provider (dark) rather than the nested provider (light):
![image](https://github.com/mantinedev/mantine/assets/115905886/31db5845-fedc-4a0f-9852-9b6a2b9ad477)

**Reply by hunterwilhelm** [reply]
For me, this is all I had to do.

```
...

const ref = useRef<HTMLElement | null>(null);

return (
  <MantineProvider
    getRootElement={() => ref.current ?? undefined}
    forceColorScheme="dark"
  >
    <section
      ref={ref}

      ...
```
(I didn't need to change the text color, just a Segmented Control)

**Reply by kevinbeal** [reply]
The problem is the CSS, not the JS. The selectors are written like this `[data-mantine-color-scheme=light] .m-8d3f4000`.

If you have defined a new root within a parent root element, you don't change the CSS. The selectors are still written the same. They will simply have  both `[data-mantine-color-scheme=light] .m-8d3f4000` and `[data-mantine-color-scheme=dark] .m-8d3f4000` targeted on the same element. 

If it works in your case it is purely by accident that the inner (dark) one was defined after the outer (light) one for the element you are looking at. As soon as you pick another component where the order is different (light after dark) your solution won't work for that component/element.

The order is different for different components and there's no reason to think the order will stay stable over time. A solution has to be built into the library itself somehow to work reliably across all components.

**Comment by Adg1** [comment]
This is the complete solution if someone is still struggling, 
```tsx
import { MantineProvider } from '@mantine/core';
import { useId } from 'react';

const ColorSchemedChildren = ({
  children,
  colorScheme,
}: {
  children: React.ReactNode;
  colorScheme: 'light' | 'dark';
}) => {
  const id = useId().replace(/:/g, '');
  const selector = `#${id}`;
  return (
    <div id={id}>
      <MantineProvider
        getRootElement={() => document.querySelector<HTMLElement>(selector)!}
        cssVariablesSelector={selector}
        forceColorScheme={colorScheme}
      >
        {children}
      </MantineProvider>
    </div>
  );
};

export const DarkSchemedChildren = ({
  children,
}: {
  children: React.ReactNode;
}) => (
  <ColorSchemedChildren colorScheme="dark">{children}</ColorSchemedChildren>
);

export const LightSchemedChildren = ({
  children,
}: {
  children: React.ReactNode;
}) => (
  <ColorSchemedChildren colorScheme="light">{children}</ColorSchemedChildren>
);
```
The extra `div` wrapper may cause some side effect in styling

**Reply by kevinbeal** [reply]
I explain above why this is not a complete solution and will fail often, even within the context of the same component. The problem is the CSS, not the JS.

**Reply by Adg1** [reply]
I tried this solution for my use case and it worked well. You're probably right that the issue might occur with MantineProvider nested at multiple levels, due to how CSS gets loaded.
Until we find a proper fix, though, this solution seems like a good temporary approach, especially for cases with a single-level provider setup.
If you could share a reproducible example of the issue you encountered, it would be helpful. I think we might be able to address it with a custom component.

**Reply by blisst** [reply]
This worked great for me when other methods didn't.

I used this to force a color scheme in a particular area of the UI, which was already inside of another, root `MantineProvider`.

Some things I found:

1. `darkHidden` and `lightHidden` won't work in the nested area because of the reason @kevinbeal is bringing up, that say your root scheme is color scheme = light and the nested scheme = dark, "[data-mantine-color-scheme=light] .m-8d3f4000 { display: none }" and "[data-mantine-color-scheme=dark] .m-8d3f4000 { display: none } " are the CSS rules that get generated and both will match your element and therefore your element will never show.

However, you can pass the computed color scheme from the root provider to the nested provider components and use that to make changes based on the root color scheme 
`const computedColorScheme = useComputedColorScheme('light', { getInitialValueInEffect: true });` 


2. If you want to have a button inside of the nested provider that changes color scheme for the ROOT provider, you need to make the `const { setColorScheme } = useMantineColorScheme();` from OUTSIDE the nested context available to your button in the nested context.  If you use `useMantineColorScheme` in the nested provider, it will pull the nested color scheme rather than the root.

**Comment by Toshinaki** [comment]
I  can see from the [doc](https://mantine.dev/theming/default-props/#default-props-with-mantinethemeprovider) that there's a `MantineThemeProvider` that also accpets a theme object.
But the doc only shows the use case for `defaultProps` for a part of the app,
so I wondering if `MantineThemeProvider` can also be used for nested themes?

**Comment by Toshinaki** [comment]
Hi @rtivital 
I just found a recent discord message: [Is it possible to have a theme and a sub-themes?](https://discord.com/channels/854810300876062770/1225745698838941736/1225746919918141511)

And you answered, as I quote
> Nested themes and MantineProviders are not currently supported

So this WAS possible but impossible now?
Any updates on this topic?

**Reply by rtivital** [reply]
This feature is not supported and was never supported in v7. I thought it would work but after checking it, I found out that it does not work.

**Reply by donverduyn** [reply]
Well obviously, the solution lies in preventing styles or. more specifically css variables from selectors matched with parent color scheme attributes, from bleeding into subtrees with nested mantine providers. As explained above, both selectors from the dark and light color scheme will be applied, so an easy solution would be to override any selectors in nested mantine providers, opposite to the current color scheme? We can override the css variables under these color schemes with inherit, as these selectors would automatically have higher specificity than those from the parent.

**Reply by donverduyn** [reply]
@rtivital what do you think about this idea, are you open to it?

**Reply by donverduyn** [reply]
We might even be able to support custom themes with this approach if we can generalize across anything that is not the current color scheme. Seems this is just a matter of setting all css variables to inherit.

**Reply by donverduyn** [reply]
a cleaner approach might be to use css layers, but i would have to figure that out.

**Reply by rtivital** [reply]
I've tried implementing that, it does not work. I'd rather not support this feature as it is not performant at all and generates a lot of (hundreds) CSS variables injected with js. To override styles for a subtree of components, use Styles API.

**Reply by donverduyn** [reply]
For my use case i revisited the problem and i can actually live without nested providers as multiple providers can still coexist in sibling trees. It might not be a solution to every case, but for things like storybook docs, it is.

**Reply by Toshinaki** [reply]
@rtivital 
sorry to comment on this old discussion.
lately I came up with an idea to do nested theme, though in another UI framework and it worked, and not tested in Mantine. 

Since most times I want nested themes only for colors and dark mode, so there's no need to re-generate all css variables, but the needed ones.
1. generating & injecting to root all css variables as the main theme
2. for a part of the layout, generate the needed css variables and inject (we have [`cssVariablesResolver`](https://mantine.dev/theming/mantine-provider/#cssvariablesresolver) but it seems to be generating all?)
   - a function that takes in an array of paths of theme object as parameter
   - e.g. cssVariablesResolver(['colors', 'black', 'white', 'headings.fontFamily'])
   - then a function to inject the generated css variables: injectCssVariables(cssVariables, target)
 
I'll test this in the next few weeks.
Do you think it's a valid solution?

**Reply by rtivital** [reply]
Not feasible to implement with the current styles soultion. I've tried it three times, there are always conflicts in style precedence order.

**Reply by Toshinaki** [reply]
Well, thanks for your hard work. Hopefully this will be fixed one day.

(I also tried on the [doc site](https://mantine.dev/) to remove the `:root` part from `:root[data-mantine-color-scheme="light"]` and `:root[data-mantine-color-scheme="dark"]`,
then add manually `data-mantine-color-scheme="dark"` to some element, and it kind of worked, 
but since some css variables are not defined (`--mantine-text-color`) it's not perfect )

**Reply by Toshinaki** [reply]
this looks great:
<img width="1913" height="880" alt="image" src="https://github.com/user-attachments/assets/f8ece8d5-6085-4c7e-b885-d304770889fe" />


and this is basically ok:
<img width="1901" height="860" alt="image" src="https://github.com/user-attachments/assets/8b6f7354-a7d9-4206-ae61-573fa48a07ca" />
